proposals/p1885.md

```diff
@@ -0,0 +1,417 @@
+`Optional` at the last elements. This guarantees that `Iterator`s cannot be used
+improperly (code safety goal), and exposes a simple interface, but does have a
+performance cost (cost of the `Optional`).
+
+If we want to retain familiarity with C++ iterators and eliminate this cost, one
+option would be to expose `Begin()` and `End()` members functions for
+`Iterable`, and allow unsafe access. The lack of pointer arithmetic might hinder
+performance improvements for unsafe use, but other solutions may exist.
+
+#### Mutable, Bidirectional, and others
+
+The current `Iterator` interfaces allows simple extension to add features such
+as bidirectional access (adding `Prev()` method), or mutable access
+(`GetMutable[addr me: Self*]() -> ElementType*`, or `Set()`)
+
+#### Chaining & Lazy evaluation
+
+A custom class extending `Iterator` can easily be implemented to be chained with
+another `Iterator` directly from the `Next()` method.
+
+```carbon
+fn Next[addr me: *Self]() -> Optional(ElementType) {
+    while (!atEnd) {
+        var item: auto = me->sourceIt->Next();
+        if (condition) {
+            return item;
+        }
+    }
+    return .Empty
+}
+```
+
+The same can be used to lazily generate or return values, and allows infinite
+generators as well.
+
+### Other interfaces
+
+#### Compatibility
+
+With `Iterable` (and `Iterator`) as the main interface, we can support other
+interfaces implicitly. For a (non yet existing) `Indexable` interface for
+example, Carbon can provide a corresponding `Iterable` parameterized final
+implementation:
+
+```carbon
+final impl forall [T:! Indexable] T as Iterable;
+```
+
+The same is applicable for users willing to automatically provide `Iterable`
+(and `for`) support for a custom interface.
+
+#### Possible interface for 'Consumable' containers
+
+Below is a suggestion for future 'consumable' containers. The `Consumable`
+interface would represent a single-use sequential access type of container.
+Items can be read (consumed) once, and are not expected to be accessible anymore
+after usage.
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+##### Range-based `for` behavior
+
+-   For each iteration, call `Consume() -> Optional(ElementType)`
+-   If the `Optional` is null, break the loop. Otherwise return the value held
+    by the Optional.
+
+##### Interface and usage
+
+```carbon
+// Carbon definition
+interface Consumable {
+    let ElementType:! Type
+    fn Consume[me: Self]() -> Optional(ElementType);
+}
+
+// Usage
+class MyContainer {
+    ...
+    impl as Consumable where .ElementType = i32 {
+        fn Consume[me: Self]() -> Optional(ElementType) {
+            ...
+        }
+    }
+}
+
+var container: MyContainer = {...};
+for (var a: auto in container) {}
+// `container` is now empty
+```
+
+### Limitations and future language improvements
```

Comment:
We should probably also mention language support for `const`, and language support for user-defined l-value expressions, because those will determine how iterators should support mutability.